### ISSUE TITLE
Unable to generate cluster.yaml from existing cluster

### ISSUE THREAD

**Issue by prashantkalkar** [kubernetes/kops#9109]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.13.0 (git-be5fb9019)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

Kubernetes Server Git version: v1.13.10
Kubectl client Git version: v1.16.0

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
I have an existing cluster setup. I tried to generate the cluster.yaml file from this existing cluster as follows:

```
kops get cluster $CLUSTER_NAME -o yaml > cluster.yaml
echo " " >> cluster.yaml
echo "---" >> cluster.yaml
echo " " >> cluster.yaml
kops get ig --name $CLUSTER_NAME -o yaml >> cluster.yaml
```

Once the yaml was created I tried to apply the same yaml I created from the cluster (without any changes to the yaml). 

```
kops replace -f cluster.yml
kops update cluster
```
**5. What happened after the commands executed?**

kops update cluster command gave out a big list of changes that are required to be applied. 

**6. What did you expect to happen?**
I was expecting that the kops upgrade should say that **no changes** are required to be applied. 
I was trying to get a cluster.yaml so that I can apply cluster changes declaratively from a CICD pipeline and persist the cluster configuration in the code (configuration as code). And later on, templatize the cluster.yaml so that it can be applied to multiple environments ensuring the consistency between them.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

I will provide the generated cluster.yaml if required as I might very well be using wrong commands to generate the yaml file. 

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
Let me know for which command I need to generate the debug logs for. There is not failiure as such in the cluster. 

**9. Anything else do we need to know?**

Let me know if there is an alternative way to generate the cluster.yaml and to templatize it.

**Comment by olemarkus** [comment]
The command to `get` the entire cluster, including instance groups, is `kops get -o yaml`.

You can read more about templating here:
https://kops.sigs.k8s.io/operations/cluster_template/

There shouldn't really be any diffs when you `get` all the resources and then `replace` them after.

**Comment by prashantkalkar** [comment]
Thanks a lot for the reply. 

Before I can try that I am facing a different issue now. It seems due to the previous run of the commands now, kops state is corrupt. 

I am getting the following error even for simple commands like `kops get ig`

`error fetching cluster "<clustername>": error reading cluster configuration "<cluster-name>": error parsing s3://<bucket-name>/<cluster-name>/config: no kind "Cluster" is registered for version "kops.k8s.io/v1alpha2" in scheme "k8s.io/kops/pkg/kopscodecs/codecs.go:33"`

This is strange since I did not run any command that should cause a state change. 
I executed the `replace` and `update` command. But `update` was not executed with `--yes` option.

Is there a way to restore my kops state to the previous version?

**Comment by olemarkus** [comment]
Are you using the latest kops release (1.16.2)? It looks like you may be using an old version of kops.

**Comment by prashantkalkar** [comment]
I am using kops version: `Version 1.13.0 (git-be5fb9019)`. Is there any possibility of restoring the state for the version I am using.

**Comment by olemarkus** [comment]
No easy way, no. It is easier to just upgrade. What is holding you back from using 1.16.2?

**Comment by prashantkalkar** [comment]
I would like to first ensure I have proper CICD setup for the current version and have a way to promote changes to all environments (including Production). But yes will plan to do the upgrade soon. 

I was able to restore the state back and was also able to generate the current YAML representing the cluster state. Adding details of what worked here: 

It seems after the initial commands I executed, the `<cluster-name>/config` file in the s3 bucket was updated with version: `apiVersion: kops.k8s.io/v1alpha2`. I had the s3 versioning enabled for the state bucket. So I compare the old value of it. The earlier value for the same was: `apiVersion: kops/v1alpha2`. 

The same has happened for all the files under the `instancegroup` path in the state bucket. 
I got this information from this thread: https://github.com/kubernetes/kops/issues/7806.

I reverted all these changes back to the original value of: `apiVersion: kops/v1alpha2`.

To generate the cluster.yaml again, I used the command suggested by you. 
`kops get -o yaml > cluster.yaml` 

And then:
```
kops replace -f cluster.yaml
kops update cluster
```
The update command did report with: `No changes need to be applied`

Thank you for all the help.
